src/main/cpp/src/histogram.cu

```diff
@@ -0,0 +1,433 @@
+    }
+  }
+
+  auto const default_mr = rmm::mr::get_current_device_resource();
+
+  // We only check if there is any row in frequencies that are negative (invalid) or zero.
+  auto check_invalid_and_zero =
+      cudf::detail::make_zeroed_device_uvector_async<int8_t>(2, stream, default_mr);
+
+  // We need to check and remember which rows are valid (positive) so we can do filtering later on.
+  auto check_valid = rmm::device_uvector<int8_t>(frequencies.size(), stream, default_mr);
+
+  thrust::for_each_n(
+      rmm::exec_policy(stream), thrust::make_counting_iterator(0), frequencies.size(),
+      [frequencies = frequencies.begin<int64_t>(), check_invalid = check_invalid_and_zero.begin(),
+       check_zero = check_invalid_and_zero.begin() + 1,
+       check_valid = check_valid.begin()] __device__(auto const idx) {
+        if (frequencies[idx] < 0) {
+          *check_invalid = 1;
+        }
+        if (frequencies[idx] == 0) {
+          *check_zero = 1;
+        }
+
+        check_valid[idx] = static_cast<int8_t>(frequencies[idx] > 0);
+      });
+
+  auto const h_checks = cudf::detail::make_std_vector_sync(check_invalid_and_zero, stream);
+  CUDF_EXPECTS(!h_checks.front(), // check invalid (negative) frequencies
+               "The input frequencies must not contain negative values.", std::invalid_argument);
+
+  auto const make_structs_histogram = [&](rmm::device_buffer &&null_mask,
+                                          cudf::size_type null_count) {
+    // Copy values and frequencies into a new structs column.
+    std::vector<std::unique_ptr<cudf::column>> values_and_frequencies;
+    values_and_frequencies.emplace_back(std::make_unique<cudf::column>(values, stream, mr));
+    values_and_frequencies.emplace_back(std::make_unique<cudf::column>(frequencies, stream, mr));
+
+    // Set null mask for the returned values.
+    // Such null mask is resulted from checking frequencies > 0.
+    if (null_count > 0) {
+      if (!values.has_nulls()) {
+        values_and_frequencies.front()->set_null_mask(std::move(null_mask), null_count);
+      } else {
+        // We need to AND the current null mask with the given null mask.
+        auto [new_null_mask, new_null_count] = cudf::detail::bitmask_and(
+            std::vector<cudf::bitmask_type const *>{
+                // Don't use values.null_mask(), to make sure no slicing.
+                values_and_frequencies.front()->view().null_mask(),
+                reinterpret_cast<cudf::bitmask_type const *>(null_mask.data())},
+            std::vector<cudf::size_type>{0, 0}, values.size(), stream, mr);
+        values_and_frequencies.front()->set_null_mask(std::move(new_null_mask), new_null_count);
+      }
+
+      // Nulls will be eventually excluded from percentile computation.
+      // However, having frequencies containing zero (corresponding to these nulls) will crash
+      // cudf MERGE_HISTOGRAM aggregation.
```

Comment:
`MERGE_HISTOGRAM` does not check for frequencies thus it is all the users' responsibility to provide positive frequencies. This is to align with cudf philosophy "garbage in garbage out".

Reply:
Note that this is similar to what we have already done in Spark plugin with NaN. We replaced NaN with nulls in a lot of operations because cudf handles NaN by different way than Spark does.